services/process-service.ts

```diff
@@ -0,0 +1,40 @@
+import * as fiberBootstrap from "../fiber-bootstrap";
+
+export class ProcessService implements IProcessService {
+	private static PROCESS_EXIT_SIGNALS = ["exit", "SIGINT", "SIGTERM"];
+	private _listeners: IListener[];
+
+	public get listenersCount(): number {
+		return this._listeners.length;
+	}
+
+	constructor() {
+		this._listeners = [];
+		_.each(ProcessService.PROCESS_EXIT_SIGNALS, (signal: string) => {
+			process.on(signal, () => this.executeAllCallbacks.apply(this));
+		});
+	}
+
+	public attachToProcessExitSignals(context: any, callback: () => IFuture<any>): void {
+		let callbackToString = callback.toString();
+
+		if (!_.some(this._listeners, (listener: IListener) => context === listener.context && callbackToString === listener.callback.toString())) {
+			this._listeners.push({ context, callback });
+		}
+	}
+
+	private executeAllCallbacks(): void {
+		fiberBootstrap.run(() => {
+			_.each(this._listeners, (listener: IListener) => {
+				listener.callback.apply(listener.context).wait();
+			});
+		});
+	}
+}
+
+interface IListener {
+	context: any;
+	callback: () => IFuture<any>;
```

Comment:
why the return type is `IFuture<any>` - we cannot execute async methods here, so let's make it void and remove the fiberBootstrap above